File: kattis/classy/classy.py
from functools import cmp_to_key

t = int(input())

SEP = "=============================="

def compare(a, b):
    a_rank = a[1]
    b_rank = b[1]
    for i in range(len(a_rank)):
        aa = a_rank[i]
        bb = b_rank[i]
        if aa != bb:
            if aa == "upper":
                return -1
            elif aa == "middle" and bb == "upper":
                return 1
            elif aa == "middle" and bb == "lower":
                    return -1
            elif aa == "lower":
                return 1
    if a[0] < b[0]:
        return -1
    elif a[0] > b[0]:
        return 1
    else:
        return 0 
            
for _ in range(t):
    n = int(input())
    arr = []
    for i in range(n):
        tokens = input().split()
        name = tokens[0][:-1]
        rank = tokens[1].split('-')[::-1]
        for j in range(10-len(rank)):
            rank.append("middle")

        arr.append((name, rank))

    arr.sort(key=cmp_to_key(compare))
    for name, rank in arr:
        print(name)
    print(SEP)

File: kattis/classy/classy.cpp
#include <iostream>
#include <vector>
#include <sstream>
#include <algorithm>

using namespace std;

const string SEP = "==============================";

struct Person {
    string name;
    vector<string> rank;
};

bool cmp(const Person& a, const Person& b) {
    vector<string> aRank = a.rank;
    vector<string> bRank = b.rank;
    for (int i = 0; i < aRank.size(); i++) {
        if (aRank[i] != bRank[i]) {
            if (aRank[i] == "upper") {
                return true;
            }
            else if (aRank[i] == "middle") {
                return (bRank[i] != "upper");
            }
            else if (aRank[i] == "lower") {
                return false;
            }
        }
    }
    return lexicographical_compare(a.name.begin(), a.name.end(), b.name.begin(), b.name.end());
}

int main() {
    string tmp;
    int t;
    cin >> t;
    for (int _ = 0; _ < t; _++) {
        int n;
        cin >> n;
        getline(cin, tmp);
        vector<Person> arr{};

        // read in data
        for (int i = 0; i < n; i++) {
            string line, name, rank, token;
            vector<string> tokens{};
            vector<string> rrank{};
            getline(cin, line);
            istringstream ss(line);
            while (getline(ss, token, ' ')) {
                tokens.push_back(token);
            }
            name = tokens[0];
            rank = tokens[1];
            name.pop_back();
            istringstream ss2(rank);
            while (getline(ss2, token, '-')) {
                rrank.push_back(token);
            }
            reverse(rrank.begin(), rrank.end());
            arr.push_back(Person{ name, rrank });
        }

        // make all rank the same length by appending "middle"
        for (auto &a: arr) {
            int s = a.rank.size();
            if (a.rank.size() < 10) {
                for (int j = 0; j < 10-s; j++) {
                    a.rank.push_back("middle");
                }
            }
        }

        // sort, highest rank first
        sort(arr.begin(), arr.end(), cmp);
        for (auto &a: arr) {
            cout << a.name << endl;
        }

        cout << SEP << endl;
    }
}